### PR TITLE
Table usability improvements

### LINKED ISSUE #3199: Usability improvements to Table View
This issue if the second (and optional) half of a series improvements to the table view UI. These improvements are divided in two groups:

1. **Styling** improvements, (issue #3198) which should be done before any other table view related additions; and
2. **Usability** improvements, (this issue) which can be done along with those updates, or later on, depending on timelines and availability.

### Usability improvements
The following updates involve small new functionalities and minor additions to make the experience more user friendly. These are not as urgent as the accessibility improvements, and most are optional, depending on what we have time for. Many can also be postponed, in which case they will be recorded in the Notion idea repository. See list below.
TableviewNEW

**Making column headers clearer and more useful**
- [ ] Include an arrow with a dropdown menu for easier to understand actions (current sorting is not very intuitive). Note that a sorted column would stay highlighted and include the icon of the type of sorting. The dropdown would also allow you to cancel the sorting. See animation:
column sorting

**Include more navigation options on top.**
- [ ] While 'filter' might be something more challenging for a later occasion, we could start with a 'hide columns' functionality. This could also be accessible from the column header dropdown described before.
- [ ] Note that in order to do this, 'number of results' and 'rows displayed' should be removed from where they are and included in the table itself.
hide fields

Additional details and specs can be found [in the following Figma file](https://www.figma.com/file/vbHVtRkdit3dqd5y8JFfXF/Table-view-redesign?node-id=0%3A1).

### LINKED ISSUE #3372: Single Submission modal should be bigger
## Description

Quoting @tinok :

> I'm reviewing applicants with the single submission view on a laptop and find the modal size could be bigger. I changed it locally to 95% height and 80% width which uses most of the screen space available:

submission modal

### PR DESCRIPTION
**Pull request by magicznyleszek**
## Related issues

Also contains code from these PRs:
- #3327 
- #3341
- #3299

Fixes #3199 
Fixes #3372

### CODE REVIEW

File: jsapp/js/components/submissions/table.es6 [hunk 1]
```diff
@@ -374,69 +482,64 @@ export class DataTable extends React.Component {
...
     let showHXLTags = this.state.showHXLTags;
-    let settings = this.props.asset.settings;
     let translationIndex = this.state.translationIndex;
-    let maxPageRes = Math.min(this.state.pageSize, this.state.tableData.length);
-    let _this = this;
+    let maxPageRes = Math.min(this.state.pageSize, this.state.submissions.length);
 
-    if (settings['data-table'] && settings['data-table']['translation-index'] !== null) {
-      translationIndex = settings['data-table']['translation-index'];
-      showLabels = translationIndex > -1 ? true : false;
```
Comment (duvld) [inline jsapp/js/components/submissions/table.es6:400]: Let me know if you can reproduce it, but for me `showLabels` is false by default. If you make a fresh form with media questions, make some submissions, then go to the data table, the XML names are being used instead

I don't know if this seemingly redundant line is responsible :D

Comment (duvld) [reply jsapp/js/components/submissions/table.es6:400]: Clicking "Save" without doing anything in the table settings fixes this

**Review by duvld — CHANGES_REQUESTED** [review]

File: jsapp/js/components/common/koboDropdown.es6 [hunk 2]
```diff
@@ -132,7 +145,7 @@ export default class KoboDropdown extends React.Component {
     let loopEl = evt.target;
     // If we check too much parents going upward looking for dropdown then we
     // assume it is not a dropdown.
-    let parentsLookupLimit = 6;
+    let parentsLookupLimit = 10;
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/common/koboDropdown.es6:148]: Oh, I did copy this function from other project, and it turns out our `React` is making quite a lot of nested nodes, so this wasn't working in some cases

File: jsapp/js/components/submissions/table.es6 [hunk 3]
```diff
@@ -66,25 +79,93 @@ export class DataTable extends React.Component {
...
+  }
+
+  /**
+   * This triggers only when asset with `content` was loaded.
+   */
+  whenLoaded() {
+    this.setState({isInitialized: true});
+  }
+
+  componentDidUpdate(prevProps) {
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/submissions/table.es6:123]: I think `componentDidUpdate` is not deprecated (it's still at https://reactjs.org/docs/react-component.html#componentdidupdate without any note), so no :)

File: jsapp/js/components/submissions/table.es6 (hunk 1 above)
Comment (magicznyleszek) [reply jsapp/js/components/submissions/table.es6:400]: I tried reproducing it:

1. Make form with some questions (both with and without media questions)
2. Make submissions
3. Open Table View

I see the Labels, not XML names. Also if I open Display Options I see "Labels" selected.

Comment (magicznyleszek) [reply jsapp/js/components/submissions/table.es6:400]: Could you try reproducing it again? Maybe the form needs to have more languages?

File: jsapp/js/components/submissions/table.es6 [hunk 4]
```diff
@@ -578,39 +710,48 @@ export class DataTable extends React.Component {
...
-            return row.value;
+            return (<span className='trimmed-text'>{row.value}</span>);
           }
         },
       });
 
     });
 
-    columns.sort(function (a, b) {
+    columnsToRender.sort(function (a, b) {
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/submissions/table.es6]: Fixed!

File: jsapp/js/components/submissions/columnsHideForm.es6 [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      let fuse = new Fuse(this.state.allColumns, {...FUSE_OPTIONS, keys: ['fieldId', 'label']});
+      return fuse.search(this.state.filterPhrase);
+    }
+    return this.state.allColumns;
+  }
+
+  render() {
+    const filteredFieldsList = this.getFilteredFieldsList();
+    return (
+      <section className='columns-hide-form'>
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/components/submissions/columnsHideForm.es6]: I don't know why I didn't, fixed :D

File: jsapp/js/actions/submissions.es6 [hunk 6]
```diff
@@ -7,12 +7,30 @@ import {dataInterface} from 'js/dataInterface';
...
 const submissionsActions = Reflux.createActions({
+  getSubmissions: {children: ['completed', 'failed']},
   bulkDeleteStatus: {children: ['completed', 'failed']},
   bulkPatchStatus: {children: ['completed', 'failed']},
   bulkPatchValues: {children: ['completed', 'failed']},
   bulkDelete: {children: ['completed', 'failed']},
 });
 
+submissionsActions.getSubmissions.listen((options) => {
+  dataInterface.getSubmissions(
```
[the remark this answers is not in the document]
Comment (magicznyleszek) [reply jsapp/js/actions/submissions.es6:36]: Good idea

File: jsapp/js/components/submissions/table.es6 (hunk 1 above)
Comment (magicznyleszek) [reply jsapp/js/components/submissions/table.es6:400]: Found it and fixed it :)

**Review by duvld — APPROVED** [review]

**Pull request merged**